modules/javafx.graphics/src/main/java/com/sun/prism/d3d/D3DPipeline.java

```diff
@@ -207,7 +237,15 @@ public int getAdapterOrdinal(Screen screen) {
     }
 
     private static D3DResourceFactory findDefaultResourceFactory(List<Screen> screens) {
-        for (int adapter = 0, n = nGetAdapterCount(); adapter != n; ++adapter) {
+        // KCR: maybe revert this change, since I don't think it is needed
+        int adapterCount = nGetAdapterCount();
+        if (adapterCount == 0) {
+            // adapters lost, recreate
+            System.err.println("KCR: adapters lost, recreate -- will this ever happen??");
+            D3DPipeline.getInstance().reinitialize();
+            return null;
```

Comment:
I have those, probably at the beginning of the reconnect to remote desktop:

> D3DContext::dispose
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> KCR: adapters lost, recreate -- will this ever happen??
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> KCR: adapters lost, recreate -- will this ever happen??
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> KCR: adapters lost, recreate -- will this ever happen??
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> constructor: com.sun.prism.d3d.D3DContext@29e6e497

Reply:
I see. Thanks for confirming. I wonder whether it will reinitialize correctly without this? Ideally we wouldn't reinitialize (with the associated destruction and recreation of the native d3d device) multiple times like this.

Reply:
Ok, I tried commenting the reinitialize here out.
That ended in "KCR: adapters lost, recreate -- will this ever happen??" on every event and no recover.

Reply:
OK, looks like it is needed then. Thanks.

Reply:
I would still like to understand this: @Schmidor does this happen every time for you? Is there something you are doing in addition to disconnecting from RDP and reconnecting (e.g., putting the remote device to sleep, unplugging an external monitor, etc)?

Reply:
Just closing the RDP connection. The errors occur on the reconnect.
The application is a WebView embedded in a JFXPanel in Swing.
Running on HighDPI / 200% on a Windows 10. So there is JFX/Swing interop and scaling involved, if that might influence that issue.

Reply:
I tried that, but still didn't reproduce it. I pushed a couple additional changes, plus some debugging code (in particular, it will give a stack trace of all calls to reinitialize).

If you get a spare minute, can you try it again and attach the log file (or send it to me offline)?

In the mean time, I'll look into the more serious issue, which is the problem with media failing to reconnect.

Reply:
[rdp-reconnect.txt](https://github.com/openjdk/jfx/files/6039065/rdp-reconnect.txt)
The reset() fails after the first try because factories is null then. I'm adding a check in the loop there and try again.

Reply:
After eliminating the NPE
[rdp-reconnect2.txt](https://github.com/openjdk/jfx/files/6039263/rdp-reconnect2.txt)

Reply:
OK. The interesting thing is:

> Error: unable to reinitialize D3D graphics device

This means that the native call to `nInit` failed for some reason. I'm interested in knowing why `nInit` fails. Can you run it with the following env variable and system property set?

```
export NWT_TRACE_LEVEL=3
java -Dprism.verbose=true ...
```

It will be pretty noisy.

Reply:
[rdp-reconnect3.txt](https://github.com/openjdk/jfx/files/6039443/rdp-reconnect3.txt)

Reply:
Thanks for doing the test run and providing the log. The following shows the error:

```
        (I) D3DPipeline_nInit
        (I) D3DPPLM::InitAdapters()
        (I) CheckAdaptersInfo
        (I) ------------------
----->  (E) D3DPPLM::CheckAdaptersInfo: no suitable adapters found
----->  (E) InitD3D: failed to init adapters
        (I) D3DPPLM::ReleaseD3D()
        (I) D3DPPLM::ReleaseAdapters()
        (W) Release: resource pd3d9=0x2d38fba0 not released: refs=1
        (I) D3DPipeline_nDispose
        ...
        D3DPipelineManager: Created D3D9Ex device
        Error: unable to reinitialize D3D graphics device
```

This is the first call to `nInit` after we get `D3DERR_DEVICEREMOVED`.

I'll dig further.